compiler/rustc_parse/src/parser/token_type.rs

```diff
@@ -0,0 +1,631 @@
+use rustc_ast::token::TokenKind;
+use rustc_span::symbol::{Symbol, kw, sym};
+
+/// Used in "expected"/"expected one of" error messages. Tokens are added here
+/// as necessary. Tokens with values (e.g. literals, identifiers) are
+/// represented by a single variant (e.g. `Literal`, `Ident`).
+///
+/// It's an awkward representation, but it's important for performance. It's a
+/// C-style parameterless enum so that `TokenTypeSet` can be a bitset. This is
+/// important because `Parser::expected_token_types` is very hot. `TokenType`
+/// used to have variants with parameters (e.g. all the keywords were in a
+/// single `Keyword` variant with a `Symbol` parameter) and
+/// `Parser::expected_token_types` was a `Vec<TokenType>` which was much slower
+/// to manipulate.
+///
+/// We really want to keep the number of variants to 128 or fewer, so that
+/// `TokenTypeSet` can be implemented with a `u128`.
```

Comment:
Suggestion: @nnethercote could you also include this comment (or even better, a short high-level description of this scheme) in the PR description, if only to make it easier to search a few years down the line (i.e. when someone wants to look up "huh, why is the `TokenType` representation in this form")?

Reply:
Um, isn't a comment on `TokenType` better for answering that question than a description in the PR that modified it?

Reply:
That's fine too.